### PR TITLE
Review WS changes

### PR DESCRIPTION
**Pull request by motorina0**
Post-merge review. See comments to code also.

Is the code inside`/nostr` still needed?

Refactor suggestions:
 - keep `views_api` for http/ws logic only. Move business logic functions (`marshall_nostr_filters`, `add_nostr_subscription`) into `services.py`

 - extract `NostrRouter`

```py
class NostrRouter:

    def __init__(self, websocket):
        self.subscriptions: List[str] = []
        self.connected: bool = True

        self.websocket = websocket

        self.tasks: List[asyncio.Task] = []

    async def client_to_nostr(self):
        ...

    async def nostr_to_client(self):
        ...

    async def start(self):
        self.tasks.append(asyncio.create_task(self.client_to_nostr()))
        self.tasks.append(asyncio.create_task(self.nostr_to_client()))

    async def stop(self):
        for t in self.tasks:
            t.cancel()
```

```py
@nostrclient_ext.websocket("/api/v1/relay")
async def ws_relay(websocket: WebSocket):
    """Relay multiplexer: one client (per endpoint) <-> multiple relays"""
    await websocket.accept()

    router = NostrRouter(websocket)
    router.start()

    while True:
        await asyncio.sleep(10)
        if not router.connected:
            for s in router.subscriptions:
                client.relay_manager.close_subscription(s)
            router.stop()
            break
```

### CODE REVIEW

File: views_api.py [hunk 1]
```diff
@@ -80,76 +73,107 @@ async def api_delete_relay(relay: Relay):  # type: ignore
...
+
+async def add_nostr_subscription(json_str):
+    """Parses a (string) request from a client. If it is a subscription (REQ), it will
+    register the subscription in the nostr client library that we're using so we can
+    receive the callbacks on it later"""
+    json_data = json.loads(json_str)
+    assert len(json_data)
+    if json_data[0] == "REQ":
+        subscription_id = json_data[1]
+        fltr = json_data[2]
```
Comment (motorina0) [inline views_api.py:104]: - why bother processing the filters, just forward them?

File: views_api.py [hunk 2]
```diff
@@ -80,76 +73,107 @@ async def api_delete_relay(relay: Relay):  # type: ignore
...
+            # registers a subscription if the input was a REQ request
+            subscription_id = await add_nostr_subscription(json_str)
+            if subscription_id:
+                my_subscriptions.append(subscription_id)
+
+            # publish data
+            client.relay_manager.publish_message(json_str)
+
+    async def nostr_to_client(websocket):
+        """Sends responses from relays back to the client. Polls the subscriptions of this client
```
Comment (motorina0) [inline views_api.py:141]: is this comment written by ChatGTP?

File: views_api.py [hunk 3]
```diff
@@ -1,29 +1,22 @@
...
-from fastapi.responses import JSONResponse
-
-from starlette.exceptions import HTTPException
-from sse_starlette.sse import EventSourceResponse
 from loguru import logger
 
 from . import nostrclient_ext
 
-from .tasks import client, received_event_queue
+from .tasks import client, received_event_queue, received_subscription_events
```
Comment (motorina0) [inline views_api.py:12]: hmm, is it ok for `views_api` to import from `tasks`?

File: views_api.py [hunk 4]
```diff
@@ -80,76 +73,107 @@ async def api_delete_relay(relay: Relay):  # type: ignore
...
+        """Sends responses from relays back to the client. Polls the subscriptions of this client
+        stored in `my_subscriptions`. Then gets all responses for this subscription id from `received_subscription_events` which
+        is filled in tasks.py. Takes one response after the other and relays it back to the client. Reconstructs
+        the reponse manually because the nostr client lib we're using can't do it."""
+        nonlocal connected
+        while True and connected:
+            for s in my_subscriptions:
+                if s in received_subscription_events:
+                    while len(received_subscription_events[s]):
+                        my_event = received_subscription_events[s].pop(0)
```
Comment (motorina0) [inline views_api.py:150]: how about non-event messages like `EOS` ([NIP15](https://github.com/nostr-protocol/nips/blob/master/15.md)):
```json
["EOSE", <subscription_id>]
```

File: views_api.py (hunk 1 above)
Comment (callebtc) [reply views_api.py:104]: That's how the library works 🤷: https://github.com/jeffthibault/python-nostr

Not important enough to justify a change of the lib imo.

File: views_api.py (hunk 3 above)
Comment (motorina0) [reply views_api.py:12]: one possibility here: https://github.com/lnbits/nostrmarket/blob/main/__init__.py#L43-L54

File: tasks.py [hunk 5]
```diff
@@ -75,9 +77,28 @@ async def subscribe_events():
...
+                eventMessage.event
+            )
+        else:
+            received_subscription_events[eventMessage.subscription_id] = [
+                eventMessage.event
+            ]
+
+        asyncio.run(received_event_queue.put(eventMessage))
 
     t = threading.Thread(
```
[the remark this answers is not in the document]
Comment (motorina0) [reply tasks.py:103]: In a similar manner with the [shceduled_tasks](https://github.com/lnbits/lnurlp/blob/main/views_api.py#L172-L180)

**Pull request closed without merging**